### PR TITLE
Add tests for CompareInfo

### PR DESCRIPTION
**Pull request by hughbe**
- Add tests that I felt were lacking in the library.
- Brings coverage up to 100%

It turns out some of my recent work was duplicated. There was already a
CompareInfoTests file, so there were two tests for each method.
I fixed this by removing the old tests, and combined the old tests into
the newer tests that are split up into multiple files

### CODE REVIEW

File: src/System.Globalization/tests/CompareInfo/CompareInfoTests.cs [hunk 1]
```diff
@@ -2,32 +2,87 @@
...
+        public void GetHashCode(string source1, CompareOptions options1, string source2, CompareOptions options2, bool expected)
+        {
+            CompareInfo invariantCompare = CultureInfo.InvariantCulture.CompareInfo;
+            Assert.Equal(expected, invariantCompare.GetHashCode(source1, options1).Equals(invariantCompare.GetHashCode(source2, options2)));
+        }
+
+        [Fact]
+        public void GetHashCode_EmptyString()
+        {
+            Assert.Equal(0, CultureInfo.InvariantCulture.CompareInfo.GetHashCode("", CompareOptions.None));
```
Comment (stephentoub) [inline src/System.Globalization/tests/CompareInfo/CompareInfoTests.cs]: This was part of a test that was previously marked as ActiveIssue on Unix.  It's causing the build to fail.  I suggest you review the CI logs to see if there are others.

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test  please

**Comment by stephentoub** [comment]
@hughbe, I get nervous any time I see such large swaths of tests getting deleted.  I know you say they're duplicates... in comments can you add links to the corresponding tests that cover the same thing that remain after this change?

Can you also separate that change into its own commit?  It's difficult to see what changes are due to that removal/redistribution vs those due to you "adding tests that you felt were lacking".

**Comment by hughbe** [comment]
Hahah good point, I understand! I'll split the commits soon

**Comment by stephentoub** [comment]
Thanks, I appreciate it.

**Comment by hughbe** [comment]
@stephentoub I've split up the tests, and added comments

**Comment by stephentoub** [comment]
> and added comments

Where?  I was expecting to see some GitHub comments on the deletions pointing to where the corresponding duplicates were.  Did I misunderstand, or am I missing them?  Thanks.

**Comment by hughbe** [comment]
Uh oh I lost them in the rebase to fix the merge conflicts

File: src/System.Globalization/tests/CompareInfo.TestData.cs [hunk 2]
```diff
@@ -1,324 +0,0 @@
...
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-using System;
-using System.Collections.Generic;
-using System.Globalization;
-
-public partial class CompareInfoTests
-{    
-    public static IEnumerable<object[]> CompareToData
```
Comment (hughbe) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: https://github.com/hughbe/corefx/blob/globalization-tests-add/src/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs#L31

File: src/System.Globalization/tests/CompareInfo.TestData.cs [hunk 3]
```diff
@@ -1,324 +0,0 @@
...
-            yield return new object[] { "", "FooBA\u0300R", "Foo\u0400B\u00C0R", -1, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "Test's", "Tests", 0, CompareOptions.IgnoreSymbols };
-            yield return new object[] { "", "Test's", "Tests", 1, CompareOptions.None };
-            yield return new object[] { "", "Test's", "Tests", -1, CompareOptions.StringSort };
-            yield return new object[] { "", new string('a', 5555), new string('a', 5555), 0, CompareOptions.None };
-            yield return new object[] { "", new string('a', 5555), new string('a', 5555) + "b", -1, CompareOptions.None };
-        }
-    }
-
-    public static IEnumerable<object[]> IndexOfData
```
Comment (hughbe) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: https://github.com/hughbe/corefx/blob/globalization-tests-add/src/System.Globalization/tests/CompareInfo/CompareInfoTests.IndexOf.cs#L19

File: src/System.Globalization/tests/CompareInfo.TestData.cs [hunk 4]
```diff
@@ -1,324 +0,0 @@
...
-            yield return new object[] { "", "FooBar", "Foo\u0400Bar", -1, CompareOptions.Ordinal };
-            yield return new object[] { "", "~FooBar", "Foo\u0400Bar", -1, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "TestFooBA\u0300R", "Foo\u0400B\u00C0R", -1, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "More Test's", "Tests", 5, CompareOptions.IgnoreSymbols };
-            yield return new object[] { "", "More Test's", "Tests", -1, CompareOptions.None };
-            yield return new object[] { "", "cbabababdbaba", "ab", 2, CompareOptions.None };
-        }
-    }
-
-    public static IEnumerable<object[]> LastIndexOfData
```
Comment (hughbe) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: https://github.com/hughbe/corefx/blob/globalization-tests-add/src/System.Globalization/tests/CompareInfo/CompareInfoTests.LastIndexOf.cs#L16

File: src/System.Globalization/tests/CompareInfo.TestData.cs [hunk 5]
```diff
@@ -1,324 +0,0 @@
...
-            yield return new object[] { "", "FooBar", "Foo\u0400Bar", -1, CompareOptions.Ordinal };
-            yield return new object[] { "", "~FooBar", "Foo\u0400Bar", -1, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "TestFooBA\u0300R", "Foo\u0400B\u00C0R", -1, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "More Test's", "Tests", 5, CompareOptions.IgnoreSymbols };
-            yield return new object[] { "", "More Test's", "Tests", -1, CompareOptions.None };
-            yield return new object[] { "", "cbabababdbaba", "ab", 10, CompareOptions.None };
-        }
-    }
-
-    public static IEnumerable<object[]> IsPrefixData
```
Comment (hughbe) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: https://github.com/hughbe/corefx/blob/globalization-tests-add/src/System.Globalization/tests/CompareInfo/CompareInfoTests.IsPrefix.cs#L16

File: src/System.Globalization/tests/CompareInfo.TestData.cs [hunk 6]
```diff
@@ -1,324 +0,0 @@
...
-            yield return new object[] { "", "FooBar", "Foo\u0400Bar", false, CompareOptions.None };
-            yield return new object[] { "", "FooBar", "Foo\u0400Bar", false, CompareOptions.Ordinal };
-            yield return new object[] { "", "FooBar", "Foo\u0400Bar", false, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "FooBA\u0300R", "Foo\u0400B\u00C0R", false, CompareOptions.IgnoreNonSpace };
-            yield return new object[] { "", "Test's can be interesting", "Tests", true, CompareOptions.IgnoreSymbols };
-            yield return new object[] { "", "Test's can be interesting", "Tests", false, CompareOptions.None };
-        }
-    }
-
-    public static IEnumerable<object[]> IsSuffixData
```
Comment (hughbe) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: https://github.com/hughbe/corefx/blob/globalization-tests-add/src/System.Globalization/tests/CompareInfo/CompareInfoTests.IsSuffix.cs#L16

**Comment by hughbe** [comment]
Uh oh, I lost the comments in the rebase to fix a merge conflict! Fixed.

**Comment by hughbe** [comment]
Rebased to fix another merge conflict

File: src/System.Globalization/tests/CompareInfo.TestData.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: Thanks

File: src/System.Globalization/tests/CompareInfo.TestData.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: thanks

File: src/System.Globalization/tests/CompareInfo.TestData.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: thanks

File: src/System.Globalization/tests/CompareInfo.TestData.cs (hunk 6 above)
Comment (stephentoub) [inline src/System.Globalization/tests/CompareInfo.TestData.cs]: thanks

**Comment by stephentoub** [comment]
LGTM. Thanks, @hughbe.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by weshaggard** [comment]
Congratulations @hughbe with this PR you have successfully doubled the total number of tests being reported in corefx, we should give you some sort of achievement for that :). We went from about 5000 to about 500000 tests being reported in this test project after this PR, which is putting a fair amount of stress on our test reporting system. So we need to look into what we are doing here a little more. 

Digging in a little I see a few things we should consider:

1) Are all these tests adding value or can they be streamlined without losing coverage?
2) Should we switch away from using Theory here and instead run the tests in one Fact in a loop? This would eliminate all the individual test result entries but still enable the coverage in a reasonable way. 
3) Slightly orthogonal but while looking at this I noticed we are generating random test data in these tests. That kind of test data generally reduces the reliability of our test runs and so we should try and avoid it. 

cc @stephentoub @Chrisboh

**Comment by hughbe** [comment]
Are you sure I've increased the number of tests by 49000! I'd be surprised if that actually happened? Are you sure there's not a bug in the reporting system?
Further, this PR removes some tests so I would've expected a net decrease. 

I can delete the random data generation if u want

**Comment by stephentoub** [comment]
> Are you sure I've increased the number of tests by 49000! I'd be surprised if that actually happened? Are you sure there's not a bug in the reporting system?

@hughbe, yes, it's coming from here https://github.com/dotnet/corefx/blob/master/src/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs#L237-L245 , and in particular that inner loop which is running from 0 to some value up to 65K, and doing it 40 times.

> Are all these tests adding value or can they be streamlined without losing coverage?

No.  I've submitted a PR to remove these random tests.

> Should we switch away from using Theory here and instead run the tests in one Fact in a loop?

No :smile:  In general theories are quite valuable, and I don't want to see us move away from them.  We should just make sure we're not running unnecessary tests, which can happen regardless of whether theories are used (e.g. the same loop that's here could have been inside of a [Fact]).  If there's an issue in being able to report on them well, we should fix that in the reporting.

> That kind of test data generally reduces the reliability of our test runs

I believe it's using a fixed seed.

File: src/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs [hunk 7]
```diff
@@ -256,20 +253,99 @@ public static IEnumerable<object[]> Compare_Random_TestData()
                 yield return new object[] { s_currentCompare, string1, string2, CompareOptions.Ordinal, PredictCompareOrdinalResult(string1, string2) };
             }
         }
-        
+
         [Theory]
-        [MemberData(nameof(Compare_TestData))]
+        [MemberData(nameof(Compare_Basic_TestData))]
+        [MemberData(nameof(Compare_Random_TestData))]
```
Comment (hughbe) [inline src/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs]: FYI this was the problematic addition (the test data already existed but wasn't enabled so I enabled it). Sorry

Comment (weshaggard) [inline src/System.Globalization/tests/CompareInfo/CompareInfoTests.Compare.cs]: @hughbe don't worry about it. We definitely appreciate all the work you have been doing in our repo.

**Pull request merged**